docs/vhost-user-net-testing.md

```diff
@@ -0,0 +1,126 @@
+             |    +----------+          +-------------+-----------+-------------+          +----------+    |           
+             |               |          |                                       |          |               |           
+             |vm1            |          |                  dpdk                 |          |           vm2 |           
+             |               |          |                                       |          |               |           
+          +--+---------------------------------------------------------------------------------------------+--+        
+          |  |                                       hugepages                                             |  |        
+          |  +---------------------------------------------------------------------------------------------+  |        
+          |                                                                                                   |        
+          |                                              host                                                 |        
+          |                                                                                                   |  
+          +---------------------------------------------------------------------------------------------------+        
+```
+## Prerequisites
+
+Prior to running the test, the following steps need to be performed, [vhost-user-ovs-dpdk](https://wiki.qemu.org/Documentation/vhost-user-ovs-dpdk) is a good reference for detailing them.
+- Enable hugepages
+- Install kernel-devel
+- Build DPDK from source
+- Build OVS from source
+
+## Test
+The test runs with multiple queue (MQ) support enabled, using 2 pairs of TX/RX queues defined for both OVS and the virtual machine. Here are the details on how the test can be run.
+
+_Setup OVS_
+
+`ovs_test.sh` is created to setup and start OVS. OVS will provide the `dpdkvhostuser` backend running in server mode.
+```bash
+mkdir -p /var/run/openvswitch
+modprobe openvswitch
+killall ovsdb-server ovs-vswitchd
+rm -f /var/run/openvswitch/vhost-user*
+rm -f /etc/openvswitch/conf.db
+export DB_SOCK=/var/run/openvswitch/db.sock
+ovsdb-tool create /etc/openvswitch/conf.db /usr/share/openvswitch/vswitch.ovsschema
+ovsdb-server --remote=punix:$DB_SOCK --remote=db:Open_vSwitch,Open_vSwitch,manager_options --pidfile --detach
+ovs-vsctl --no-wait init
+ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-lcore-mask=0xf
+ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem=1024
+ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+ovs-vsctl --no-wait set Open_vSwitch . other_config:pmd-cpu-mask=0xf
+ovs-vswitchd unix:$DB_SOCK --pidfile --detach --log-file=/var/log/openvswitch/ovs-vswitchd.log
+ovs-vsctl add-br ovsbr0 -- set bridge ovsbr0 datapath_type=netdev
+ovs-vsctl add-port ovsbr0 vhost-user1 -- set Interface vhost-user1 type=dpdkvhostuser
```

Comment:
On my Ubunutu 18.04, run into below error
the output of running ovs_test.sh

+ ovs-vsctl add-br ovsbr0 -- set bridge ovsbr0 datapath_type=netdev
+ ovs-vsctl add-port ovsbr0 vhost-user1 -- set Interface vhost-user1 type=dpdkvhostuser
ovs-vsctl: Error detected while setting up 'vhost-user1': could not add network device vhost-user1 to ofproto (Invalid argument).  See ovs-vswitchd log for details.
ovs-vsctl: The default log directory is "/var/log/openvswitch".
+ ovs-vsctl add-port ovsbr0 vhost-user2 -- set Interface vhost-user2 type=dpdkvhostuser
ovs-vsctl: Error detected while setting up 'vhost-user2': could not add network device vhost-user2 to ofproto (Invalid argument).  See ovs-vswitchd log for details.

from /var/log/openvswitch/ovs-vswitchd.log 

2020-03-03T21:20:02.902Z|00105|netdev|WARN|vhost-user2: arguments provided to device that is not configurable
2020-03-03T21:20:02.902Z|00106|dpdk|ERR|MBUF: error setting mempool handler
2020-03-03T21:20:02.902Z|00107|netdev_dpdk|ERR|Failed to create mempool "ovse503631800021580262144" with a request of 262144 mbufs
2020-03-03T21:20:02.902Z|00108|netdev_dpdk|ERR|Failed to create memory pool for netdev vhost-user2, with MTU 1500 on socket 0: Invalid argument
2020-03-03T21:20:02.902Z|00109|dpif_netdev|ERR|Failed to set interface vhost-user2 new configuration
2020-03-03T21:20:02.903Z|00110|dpdk|INFO|VHOST_CONFIG: Linear buffers requested without external buffers, disabling host segmentation offloading support

The hugepage have been enabled 
 cat /proc/meminfo | grep Huge
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
HugePages_Total:    2048
HugePages_Free:     1536
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB

The kernel cmd is set as：
cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-4.15.0-88-generic root=UUID=eb27aca8-a201-4bb3-95fb-a1b2a55b9798 ro default_hugepagesz=2M hugepagesz=2M hugepages=2048

Reply:
could you share your script to start cloud-hypervisor vm?

Reply:
I have no special script to start CLH. All the scripts are from yours. :) And I was blocked in this step, and have not reached the step to start VM.

Reply:
Ah, yes, you are still configuring OVS+DPDK environment. I have not tried ubuntu and did not encounter such issue, could you please ping me and provide your environment info if it's possible, then I can log on to have a look?

Reply:
I have no ubuntu at hand for a try :-)

Reply:
I'd like to summarize the steps I used to try it on my Ubunutu 18.04 machine
1. OS System 18.04
2. DPDK: The latest git tree git://dpdk.org/dpdk
3. OVS: the latest git tree clone https://github.com/openvswitch/ovs
(Maybe I should use some specific release of these projects.)
4. Building DPDK:  Using the step in https://wiki.qemu.org/Documentation/vhost-user-ovs-dpdk recommend by this document. I added step "ldconfig /lib64" so OVS applications can locate DPDK library.
5. Building OVS: using exactly the same steps suggested in this document 
6. Enabling hugepage on host machine: I try to enable hugebage by editing the kernel booting cmdline suggested in this document and also tried to twist the /proc/sys/vm/*_hugepage_*. I think both of them works fine
7. ovs_test.sh, using almost the same shell script. I only add line "mkdir -p /var/log/openvswitch" in the beginning of the script to create the logging file directory.

I guess the problem may come from the below aspects:
1. The release of OVS and DPDK we should use. 
2. the proper way to install of DPDK library for OVS.
Just for your reference.
I hope that can help. :)
Thanks!

Reply:
As I have mentioned, the current doc is based on redhat machine. For how to install and configure OpenVswitch-DPDK on ubuntu server, I think you could refer to this page https://help.ubuntu.com/lts/serverguide/DPDK.html

Reply:
I will update the document with redhat specifies, once you complete the ubuntu steps, we could add the ubuntu part.

Reply:
@bjzhjing please update the document by providing these two links you've been mentioning for Ubuntu machines. I think this will be enough for now.

Reply:
Add the two additional references, and update around.